### PR TITLE
Source queue

### PR DESCRIPTION
**Pull request by codehag**

### CODE REVIEW

**Comment by codehag** [comment]
try run: https://treeherder.mozilla.org/#/jobs?repo=try&revision=382ec7c5be5db7b86c2a74a6e8661603e2cc4f13

File: src/actions/sources.js [hunk 1]
```diff
@@ -114,8 +114,19 @@ export function newSources(sources: Source[]) {
...
+    dispatch({
+      type: "ADD_SOURCES",
+      sources: filteredSources
+    });
+
     for (const source of filteredSources) {
-      dispatch(newSource(source));
+      dispatch(loadSourceMap(source));
+      dispatch(checkSelectedSource(source));
+      dispatch(checkPendingBreakpoints(source.id));
```
Comment (jasonLaster) [inline src/actions/sources.js]: oops - i think `dispatch(newSource(source));` was a merge conflict...

we should probably remove `newSource` soon if we don't intend on using it.

File: src/client/firefox/events.js [hunk 2]
```diff
@@ -53,6 +59,7 @@ async function paused(_: "paused", packet: PausedPacket) {
 
   if (why.type != "alreadyPaused") {
     const pause = createPause(packet, response);
+    flushSourceQueue();
```
Comment (jasonLaster) [inline src/client/firefox/events.js]: i think i'd prefer `sourceQueue.flush()` but that's minor 

`import * as sourceQueue from '../utils/source-queue'`

File: src/utils/source-queue.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  queue();
+}
+
+export function flushSourceQueue() {
+  queue.flush();
+}
+
+export function clearSourceQueue() {
+  queue.cancel();
+}
```
Comment (jasonLaster) [inline src/utils/source-queue.js]: perhaps the queue should live in `client/utils` not sure, just seems like something we should hide a bit..

File: src/test/mochitest/browser_dbg-breaking.js [hunk 4]
```diff
@@ -13,7 +13,7 @@ add_task(async function() {
   reload(dbg);
   await waitForPaused(dbg);
 
-  await waitForLoadedSource(dbg, "doc-scripts.html");
+  await waitForSelectedSource(dbg, "doc-scripts.html");
```
Comment (jasonLaster) [inline src/test/mochitest/browser_dbg-breaking.js]: yikes... perhaps we should audit these `waitForLoadedSource`...

File: src/reducers/sources.js [hunk 5]
```diff
@@ -55,16 +55,14 @@ function update(
...
     case "ADD_SOURCES": {
-      action.sources.forEach(source => {
-        state = state.mergeIn(["sources", source.id], source);
-      });
-
-      return state;
+      return action.sources.reduce(
+        (newState, source) => updateSource(newState, source),
+        state
+      );
```
Comment (jasonLaster) [inline src/reducers/sources.js]: :+1:

File: src/test/mochitest/browser_dbg-breaking.js (hunk 4 above)
Comment (codehag) [reply src/test/mochitest/browser_dbg-breaking.js]: yikes yes you are right

**Comment by codehag** [comment]
new try run: https://treeherder.mozilla.org/#/jobs?repo=try&revision=c12a713ff7c0412b3d6bd0567a3c429d41f01d47

**Pull request merged**